### ISSUE TITLE
`dockerd` Systemd unit should be of `Type=notify` (packaging issue)

### ISSUE THREAD

**Issue by unshare** [moby/moby#37990]
**Description**
[Not sure I'm filing the issue to correct repo -- if this is not the right place, please advise.]

Docker daemon issues `sd_notify` after initialization. I have an automation script (actually, an installer) that proceeds immediately after Systemd reports `docker.service` as active. It fails if launched against Docker CE v18.09.0-beta3.

The root cause is `docker.service` being a simple service (not `Type=notify`) in recent RPM builds for CentOS.
Introduced in `f99ffa4b5ef8c85fa5fbeb8fdc95403f9a02ceaf` to `docker/docker-ce-packaging` https://github.com/docker/docker-ce-packaging/commit/f99ffa4b5ef8c85fa5fbeb8fdc95403f9a02ceaf.
Changing back `Type=simple` to `Type=notify` (I use drop-ins) fixes the issue for me.

**Steps to reproduce the issue:**
1. Update Docker CE on CentOS 7 to >=18.09.0-ce-tp6
2. Install `dbus-python`, `python-docker-py` (the one shipped with CentOS 7 Extras) packages.
2. Launch a script in "additional information"

**Describe the results you received:**
```
Stopping dockerd(8)...
Starting dockerd(8)...
Traceback (most recent call last):
  File "./repro.py", line 134, in <module>
    print 'Docker version: %s' % docker.Client().info()['ServerVersion']
  File "/usr/lib/python2.7/site-packages/docker/api/daemon.py", line 33, in info
    return self._result(self._get(self._url("/info")), True)
  File "/usr/lib/python2.7/site-packages/docker/utils/decorators.py", line 47, in inner
    return f(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/docker/client.py", line 139, in _get
    return self.get(url, **self._set_request_timeout(kwargs))
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 476, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', error(2, 'No such file or directory'))
```

**Describe the results you expected:**
```
Stopping dockerd(8)...
Starting dockerd(8)...
Docker version: 18.09.0-beta3
```

**Additional information you deem important (e.g. issue happens only occasionally):**
It's a race condition, but I have a 100% reproduction with the script attached.

The test script:
```
#!/usr/bin/env python2.7

import threading
import dbus

from xml.etree import ElementTree as Et

try: from typing import *
except ImportError: pass

from gi.repository import GObject
from dbus.mainloop.glib import DBusGMainLoop

DBusGMainLoop(set_as_default=True)

IManager = 'org.freedesktop.systemd1.Manager'
IPeer = 'org.freedesktop.DBus.Peer'
IIntrospectable = 'org.freedesktop.DBus.Introspectable'
IProperties = 'org.freedesktop.DBus.Properties'
IUnit = 'org.freedesktop.systemd1.Unit'
IService = 'org.freedesktop.systemd1.Service'


class SystemCtl(object):
    def __init__(self):
        self._loop = GObject.MainLoop()
        self._bus = dbus.SystemBus(private=True)
        """:type: dbus._dbus.SystemBus"""
        self._mgr = dbus.Interface(
            self._bus.get_object('org.freedesktop.systemd1', '/org/freedesktop/systemd1'), IManager)
        self._lock = threading.Lock()

    def __del__(self): pass

    def reset_failed(self):
        # type: () -> None
        self._mgr.ResetFailed()

    def get_unit(self, unit_name):
        # type: (str) -> Union[_Unit, _Service]
        obj = self._bus.get_object('org.freedesktop.systemd1', self._mgr.GetUnit(unit_name))
        xml = Et.fromstring(dbus.Interface(obj, IIntrospectable).Introspect())  # type: Et.Element
        iface_names = [i.attrib['name'] for i in xml.findall('./interface')]
        properties = dbus.Interface(obj, IProperties)
        unit = dbus.Interface(obj, IUnit)
        if IService in iface_names:
            return _Service(self, unit_name, properties, unit)
        return _Unit(self, unit_name, properties, unit)


class _Job(object):
    def __init__(self, systemctl, job):
        # type: (SystemCtl, dbus.Interface) -> None
        self._systemctl = systemctl
        self._job = job
        self._result = None

    def wait_for_removal(self, timeout=15000):
        # type: (int) -> str
        # noinspection PyProtectedMember
        lock, mgr, loop = self._systemctl._lock, self._systemctl._mgr, self._systemctl._loop
        if not lock.acquire(False): raise EnvironmentError('not re-entrant')
        try:
            src = GObject.timeout_add(timeout, loop.quit)
            try:
                mgr.Subscribe()
                try:
                    self._result = None

                    def _on_job_removed(_1, j, _2, r):
                        if j == self._job:
                            self._result = r
                            loop.quit()

                    mgr.connect_to_signal('JobRemoved', _on_job_removed)
                    loop.run()

                    return self._result or 'timeout'
                finally: mgr.Unsubscribe()
            finally: GObject.source_remove(src)
        finally: lock.release()


class _Unit(object):
    def __init__(self, systemctl, unit_name, properties, unit):
        # type: (SystemCtl, str, dbus.Interface, dbus.Interface) -> None
        self._systemctl = systemctl
        self._name = unit_name
        self._properties = properties
        self._unit = unit

    def start(self): return _Job(self._systemctl, self._unit.Start('fail'))

    def stop(self): return _Job(self._systemctl, self._unit.Stop('fail'))

    def reset_failed(self):
        # type: () -> None
        self._unit.ResetFailed()

    def unit_property(self, prop): return self._properties.Get(IUnit, prop)


class _Service(_Unit):

    def service_property(self, prop): return self._properties.Get(IService, prop)


systemctl = SystemCtl()

dockerd_unit_name = 'docker.service'
docker_service = systemctl.get_unit(dockerd_unit_name)


def _stop_dockerd():
    print('Stopping dockerd(8)...')
    docker_service.reset_failed()
    if docker_service.unit_property('ActiveState') != 'inactive':
        assert docker_service.stop().wait_for_removal() in ('done', 'timeout') \
               and docker_service.unit_property('ActiveState') == 'inactive'


def _start_dockerd():
    print('Starting dockerd(8)...')
    if docker_service.unit_property('ActiveState') != 'active':
        docker_service.reset_failed()
        assert docker_service.start().wait_for_removal() in ('done', 'timeout') \
            and docker_service.unit_property('ActiveState') == 'active'


if __name__ == '__main__':
    import docker
    _stop_dockerd()
    _start_dockerd()
    print 'Docker version: %s' % docker.Client().info()['ServerVersion']
```

**Output of `docker version`:**

```
Client:
 Version:           18.09.0-beta3
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        7718f80
 Built:             Thu Oct  4 23:53:03 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0-beta3
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       7718f80
  Built:            Thu Oct  4 23:24:09 2018
  OS/Arch:          linux/amd64
  Experimental:     true
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.09.0-beta3
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: active
 NodeID: jix520mr6i6t04g7m2hpmkr10
 Is Manager: true
 ClusterID: 2d1z0x7uh4zcup5kbb5lks0eu
 Managers: 1
 Nodes: 1
 Default Address Pool: 10.0.0.0/8
 SubnetSize: 24
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 10
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 172.23.251.242
 Manager Addresses:
  172.23.251.242:2377
Runtimes: runc kalevala
Default Runtime: kalevala
Init Binary: docker-init
containerd version: 99fc40fd6088baebe3c18b8f8093cd50d58815d6 (expected: 468a545b9edcd5932818eb9de8e72413e616e86e)
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871 (expected: 69663f0bd4b60df09991c08812a60108003fa340)
init version: fec3683
Security Options:
 seccomp
 userns
Kernel Version: 3.10.0-862.14.4.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 5.795GiB
Name: kalevala.i.lps.id-vz.net
ID: HJHJ:ATJ5:TA3Z:7HP3:SZUZ:YTOT:IT5P:VTJ4:S272:FJGI:B6K7:ET5S
Docker Root Dir: /var/lib/docker/65536.65536
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 37
 Goroutines: 153
 System Time: 2018-10-08T18:22:15.790499107Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
 provider=generic
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by thaJeztah** [comment]
For reference, the `Type=Notify` was originally added in https://github.com/moby/moby/commit/d3e5179c291a7646c71f1ca608d6700026756f7c (https://github.com/moby/moby/pull/15119)

**Comment by unshare** [comment]
Resolved with docker/docker-ce-packaging#249
